subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java

```diff
@@ -170,15 +178,18 @@ private static class DefaultCrossBuildInMemoryCache<K, V> extends AbstractCrossB
         private final Set<V> valuesForPreviousSession = new HashSet<>();
```

Comment:
🤔 this set isn't thread-safe. I see that it is only accessed from `retainValuesFromCurrentSession` and `discardRetainedValues`, but I'd add some kind of synchronization just in case - it isn't obvious if the cache instance is going to be safely published every time we may want to call these methods. Even if it is now, it is pretty far away to be accidentally broken by some unrelated refactoring.

Reply:
> I'd add some kind of synchronization just in case

Done.